File: src/sc_neurocore/accel/julia/synapses/short_term_plasticity.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia acceleration for synapses/short_term_plasticity

module ShortTermPlasticityAccel

using Statistics, LinearAlgebra

mutable struct ShortTermPlasticitySynapseState
    x::Float64
    u::Float64
    u_base::Float64
    tau_d::Float64
    tau_f::Float64
    amplitude::Float64
    dt::Float64
end

function ShortTermPlasticitySynapseState()
    ShortTermPlasticitySynapseState(1.0, 0.5, 0.5, 200.0, 20.0, 1.0, 1.0)
end

function new_depressing(s::ShortTermPlasticitySynapseState)
    return cls(
        x=1.0,
        u=0.5,
        u_base=0.5,
        tau_d=200.0,
        tau_f=20.0,
        amplitude=1.0,
    )
end

function new_facilitating(s::ShortTermPlasticitySynapseState)
    return cls(
        x=1.0,
        u=0.1,
        u_base=0.1,
        tau_d=50.0,
        tau_f=500.0,
        amplitude=1.0,
    )
end

function step(s::ShortTermPlasticitySynapseState, pre_spike)
    # Recover between spikes.
    s.x += (1.0 - s.x) / s.tau_d * s.dt
    s.u += (s.u_base - s.u) / s.tau_f * s.dt
    if pre_spike
        # Facilitation: increase release probability.
        s.u += s.u_base * (1.0 - s.u)
        # Compute PSC before depression.
        psc = s.amplitude * s.u * s.x
        # Depression: consume resources.
        s.x -= s.u * s.x
        s.x = max(s.x, 0.0)
        return psc
    return 0.0
end

function reset(s::ShortTermPlasticitySynapseState)
    s.x = 1.0
    s.u = s.u_base
end

end # module ShortTermPlasticityAccel

File: src/sc_neurocore/accel/rust/safety/short_term_plasticity.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for short_term_plasticity

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct ShortTermPlasticitySynapse {
    pub x: f64,
    pub u: f64,
    pub u_base: f64,
    pub tau_d: f64,
    pub tau_f: f64,
    pub amplitude: f64,
    pub dt: f64,
}

impl ShortTermPlasticitySynapse {
    pub fn new() -> Self {
        Self {
            x: 1.0_f64,
            u: 0.5_f64,
            u_base: 0.5_f64,
            tau_d: 200.0_f64,
            tau_f: 20.0_f64,
            amplitude: 1.0_f64,
            dt: 1.0_f64,
        }
    }

    pub fn new_depressing(&self, ) -> f64 {
        // return cls(
        // x=1.0,
        // u=0.5,
        // u_base=0.5,
        // tau_d=200.0,
        // tau_f=20.0,
        // amplitude=1.0,
        // )
        0.0
    }

    pub fn new_facilitating(&self, ) -> f64 {
        // return cls(
        // x=1.0,
        // u=0.1,
        // u_base=0.1,
        // tau_d=50.0,
        // tau_f=500.0,
        // amplitude=1.0,
        // )
        0.0
    }

    pub fn step(&mut self, i_ext: f64) -> i32 {
        // # Recover between spikes.
        // self.x += (1.0 - self.x) / self.tau_d * self.dt
        // self.u += (self.u_base - self.u) / self.tau_f * self.dt
        // if pre_spike:
        // # Facilitation: increase release probability.
        // self.u += self.u_base * (1.0 - self.u)
        // # Compute PSC before depression.
        // psc = self.amplitude * self.u * self.x
        // # Depression: consume resources.
        // self.x -= self.u * self.x
        // self.x = max(self.x, 0.0)
        // return psc
        // return 0.0
        0 // spike indicator
    }

    pub fn reset(&mut self) {
        // self.x = 1.0
        // self.u = self.u_base
        self.x = 1.0_f64;
        self.u = 0.5_f64;
        self.u_base = 0.5_f64;
        self.tau_d = 200.0_f64;
        self.tau_f = 20.0_f64;
    }

}

pub fn validate_short_term_plasticity(state: &ShortTermPlasticitySynapse) -> bool {
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_short_term_plasticity_new() {
        let state = ShortTermPlasticitySynapse::new();
        assert!(validate_short_term_plasticity(&state));
    }

    #[test]
    fn test_short_term_plasticity_step() {
        let mut state = ShortTermPlasticitySynapse::new();
        let spike = state.step(10.0);
        assert!(spike == 0 || spike == 1);
    }
}
